### ISSUE TITLE
deprecate twisted.python.filepath.FilePath.statinfo

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4450]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4450|
|Type|enhancement|
|Created|2010-05-23 09:17:43Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-filepath-statinfo-4450-3|

This attribute was an accidentally exposed implementation detail.  Nobody should be using it.  Let's make it private.

Attachments:

* [statinfo.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0ef/0efbac5d2858bdb001322b81c2a57445af6d0367/145a3dfa9e39aea96215865f910bc21b280349c4.patch) (7272 bytes) - added by vasuman on 2013-04-16 22:02:48Z - changed variable statinfo to _statinfo; added a DeprecationWarning when variable is accessed.
* [statinfo.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0ef/0efbac5d2858bdb001322b81c2a57445af6d0367/f6ceb1e649cf2dd1c63755a0914d02615adeb1db.patch) (9204 bytes) - added by vasuman on 2013-04-17 07:53:30Z - Sorry, forgot to add unit tests and made the patch more Twisted compliant I guess

<details><summary>Searchable metadata</summary>

```
trac-id__4450 4450
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_deprecate_filepath_statinfo_4450_3 branches/deprecate-filepath-statinfo-4450-3
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1274606263000000 1274606263000000
changetime__1405490689376999 1405490689376999
version__None None
owner__cyli cyli
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @cyli|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

(In [30170]) Branching to 'deprecate-filepath-statinfo-4450'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

(In [30171]) Deprecate twisted.python.filepath.FilePath.statinfo.

Refs [#4450](https://github.com/twisted/twisted/issues/4450)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-filepath-statinfo-4450)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

(In [30172]) Fix the docstring on FilePath.statinfo

Refs [#4450](https://github.com/twisted/twisted/issues/4450)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli removed owner|
|-|-|

Fixed the epytext bug that caused the api documentation builder to fail.  [new results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-filepath-statinfo-4450&num_builds=3).

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

Filed [#4711](https://github.com/twisted/twisted/issues/4711) to remove usages of FilePath.statinfo in Twisted

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

Filed [#4712](https://github.com/twisted/twisted/issues/4712) because certain keys in statinfo do not have accessors provided by FilePath.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @cyli|
|-|-|

Thanks for working on this, cyli.

statinfo is used in two places:

 * ```twisted/protocols/ftp.py:        return [getattr(self, '_stat_' + k)(filePath.statinfo) for k in keys]```
 * ```twisted/web2/static.py:        st = self.fp.statinfo```

I don't know what our policy is on when code using something deprecated gets converted to an alternative, so I'll leave that up to you.

```test_deprecateStatinfo``` also needs a docstring.

Other than that, looks great, please merge!

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|michaelnt commented|
|-|-|

t.p.ftp.FTPAnonymousShell.list is a public method that excepts a set of keys which should retrieved from statinfo so it seems that the functionality of statinfo in filepath is needed here.

Is the reason this is being deprecated because statinfo might be None, as suggested by the doc string?

If this is the case a __getattr__ method could call self.restat when self.statinfo is None .

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

Note: there have been no comments on this ticket for a while, because there were a bunch of intermediate tickets on removing all calls to statinfo in twisted (like in FTP).

**Comment by twisted-trac** [comment]
|<img alt="vasuman's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|vasuman set owner to vasuman<br>vasuman set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="vasuman's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|vasuman set owner to @cyli<br>vasuman set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj set owner to @cyli<br>@wallrj set status to `new`|
|-|-|

Hi vasuman,

Thanks for your patch, but the deprecation had already been implemented in log:branches/deprecate-filepath-statinfo-4450 and it had already been reviewed and was waiting for some dependency tickets to be resolved (although the exact tickets haven't been listed).

See the comments from cyli and jesstess above.

Unfortunately that branch no longer merges cleanly to trunk.

```
[richard@zorin Twisted]$ unbranch Twisted
~/projects/Twisted/trunk$ svn status --quiet
C: 
~/projects/Twisted/trunk$ svn up
C: Updating '.':
C: A    twisted/topfiles/5586.doc
C: A    twisted/web/topfiles/6503.misc
C: U    doc/core/howto/application.xhtml
C: U    doc/web/howto/web-in-60/error-handling.xhtml
C: Updated to revision 38404.
~/projects/Twisted/trunk$ svn merge --non-interactive /home/richard/projects/Twisted/branches/deprecate-filepath-statinfo-4450/@30170 /home/richard/projects/Twisted/branches/deprecate-filepath-statinfo-4450/@HEAD
C: --- Merging r30171 through r38404 into '.':
C: C    twisted/test/test_paths.py
C: C    twisted/python/filepath.py
C: A    twisted/topfiles/4450.removal
C: Summary of conflicts:
C:   Text conflicts: 2
```

Cyli (or someone with SVN commit access) needs to:
 1. merge forward
 2. Change the "deprecated since" version to 13.1.0
 3. link to the tickets which you were waiting to be closed (create them if necessary) 
 4. replace has_key in the changed code.
 5. Post some clean build results

Then I agree with jesstess, that this branch is ready to be merged.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

(In [42361]) Branching to deprecate-filepath-statinfo-4450-2.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_17"></a>|@cyli removed owner|
|-|-|

Ok!  Apologies for not getting back to this for so long.  I merged forward and resolved some conflicts, and added some more tests to make sure statinfo can still be get/set, but that both getting/setting will trigger deprecation warnings.

I've updated [#4711](https://github.com/twisted/twisted/issues/4711) (the dependent ticket)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set owner to @cyli|
|-|-|

You might have to resolve some conflicts with the recent cleanup branch, but otherwise this branch looks great and I look forward to deleting `statinfo`.  Please land.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_19"></a>|@cyli commented|
|-|-|

(In [42872]) Branching to deprecate-filepath-statinfo-4450-3.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_20"></a>|@cyli commented|
|-|-|

Thanks for the review!

Merged forward, buildbot build here:  https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-filepath-statinfo-4450-3

Going to merge to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_21"></a>|@cyli set status to `closed`|
|-|-|

(In [42875]) Merge deprecate-filepath-statinfo-4450-3: Deprecate filepath.statinfo

Author: cyli
Reviewers: jesstess, rwall, glyph
Fixes: [#4450](https://github.com/twisted/twisted/issues/4450)

Deprecate filepath.statinfo in favor of accessor properties instead.
